### PR TITLE
Introduce preprocessor

### PR DESCRIPTION
**Pull request by yuliasherman**
Initial implementation of pre-execution related functionality in concord-bft.
- VB-2199 Change Replica.hpp API to expose the required information to PreProcessor.
- VB-2194 Add pre-execution specific functionality to the MsgsCommunicator module.
- VB-2196 Define new PreProcessRequest and PreProcessResult messages.
- VB-2198 Implement PreProcessor in primary replica - client requests handling.

### CODE REVIEW

**Comment by teoparvanov** [comment]
Hi Yulia, the changes look good to me at first glance. I just have one question, for my understanding:

The part where we compare the f+1 hashes and append the pre-processed request to the replica queue isn't implemented yet, right?

File: bftengine/include/bftengine/Replica.hpp [hunk 1]
```diff
@@ -25,7 +25,12 @@
 namespace bftEngine {
 
 // Possible values for 'flags' parameter
-enum MsgFlag : uint8_t { EMPTY_FLAGS = 0x0, READ_ONLY_FLAG = 0x1, PRE_EXECUTE_FLAG = 0x2, PRE_EXECUTED_FLAG = 0x4 };
+enum MsgFlag : uint8_t {
+  EMPTY_FLAGS = 0x0,
+  READ_ONLY_FLAG = 0x1,
+  PRE_PROCESS_FLAG = 0x2,
+  HAS_PRE_PROCESSED_FLAG = 0x4
```
Comment (teoparvanov) [inline bftengine/include/bftengine/Replica.hpp]: Why are we renaming those? For me, the preprocessor is a component (an implementation detail) of pre-execution. 

IMO those flags should continue to be called PRE_EXECUTE*.

**Comment by yuliasherman** [comment]
> Hi Yulia, the changes look good to me at first glance. I just have one question, for my understanding:
> 
> The part where we compare the f+1 hashes and append the pre-processed request to the replica queue isn't implemented yet, right?

You are right, I've implemented only the first part of PreProcessor for a primary replica so far.

File: bftengine/include/bftengine/Replica.hpp (hunk 1 above)
Comment (yuliasherman) [reply bftengine/include/bftengine/Replica.hpp]: Guy proposed to give a more generic name to the code related to the pre-execution feature and I agreed.

**Comment by yuliasherman** [comment]
Approval?

**Comment by teoparvanov** [comment]
Thanks Yulia - such a clean implementation!

**Comment by yuliasherman** [comment]
Thanks, have a rest, you are not expected to work today!

**Pull request merged**